### PR TITLE
fs/vfs/fs_poll.c: Only enter critical section when actually setting r…

### PR DESCRIPTION
**Pull request by jlaitine**
…events

This is an optimization.

I just got this idea, so thought that I'd write it into a pr....

### CODE REVIEW

**Comment by t-salminen** [comment]
Probably POLLALWAYS should be checked earlier

**Review by pussuw — APPROVED** [review]
Yes, this makes sense

File: fs/vfs/fs_poll.c [hunk 1]
```diff
@@ -284,23 +285,27 @@ void poll_notify(FAR struct pollfd **afds, int nfds, pollevent_t eventset)
...
+
+                  fds->revents &= ~POLLOUT;
+                }
 
-          leave_critical_section(flags);
+              leave_critical_section(flags);
+            }
 
-          if ((fds->revents != 0 || (fds->events & POLLALWAYS) != 0) &&
+          if (revents != 0 || (fds->events & POLLALWAYS) != 0 &&
```
Comment (pussuw) [inline fs/vfs/fs_poll.c:308]: This is not equivalent to the original condition, now if fds->cb = NULL you get a crash

File: fs/vfs/fs_poll.c [hunk 2]
```diff
@@ -284,24 +285,28 @@ void poll_notify(FAR struct pollfd **afds, int nfds, pollevent_t eventset)
...
+                }
 
-          leave_critical_section(flags);
+              leave_critical_section(flags);
+            }
 
-          if ((fds->revents != 0 || (fds->events & POLLALWAYS) != 0) &&
-              fds->cb != NULL)
+          if (fds->cb != NULL &&
+              (revents != 0 || (fds->events & POLLALWAYS))
```
Comment (t-salminen) [inline fs/vfs/fs_poll.c:309]: If section above might modify fds->revents value so it's not anymore same as local revents

**Review by t-salminen — APPROVED** [review]

File: fs/vfs/fs_poll.c (hunk 2 above)
Comment (jlaitine) [reply fs/vfs/fs_poll.c:310]: Yes. So you don't want to call the callback for any event that was not notified for this call, so checking fds->revents is wrong here in any case. But, if there is indeed a case that someone calls this with POLLERR and POLLOUT or POLLHUP and POLLOUT, then the POLLOUT should not be counted in. I don't think that makes much sense though (?) I can of course clear out the POLLOUT flag also from revents.

**Review by pussuw — APPROVED** [review]
LGTM at least

**Pull request merged**